source/Calamari/Deployment/PackageRetention/Model/Journal.cs

```diff
@@ -0,0 +1,102 @@
+            this.log = log;
+        }
+
+        public void RegisterPackageUse(string packageID, string version, string serverTaskID)
+        {
+            RegisterPackageUse(new PackageIdentity(packageID, version), new ServerTaskID(serverTaskID));
+        }
+
+        public void RegisterPackageUse(IVariables variables)
+        {
+            RegisterPackageUse(new PackageIdentity(variables), new ServerTaskID(variables));
+        }
+
+        public void RegisterPackageUse(PackageIdentity package, ServerTaskID serverTaskID)
+        {
+            try
+            {
+                var repository = repositoryFactory.CreateJournalRepository();
+
+                if (repository.TryGetJournalEntry(package, out var entry))
+                {
+                    entry.PackageUsage.AddUsage(serverTaskID);
+                    entry.PackageLocks.AddLock(serverTaskID);
+                }
+                else
+                {
+                    entry = new JournalEntry(package);
+                    entry.PackageUsage.AddUsage(serverTaskID);
+                    entry.PackageLocks.AddLock(serverTaskID);
+                    repository.AddJournalEntry(entry);
+                }
+
+#if DEBUG
+                log.Verbose($"Registered package use/lock for {package} and task {serverTaskID}");
+#endif
+
+                repository.Commit();
+            }
+            catch (Exception ex)
+            {
+                //We need to ensure that an issue with the journal doesn't interfere with the deployment.
+                log.Error($"Unable to register package use for retention.{Environment.NewLine}{ex.ToString()}");
+            }
+        }
+
+        public void DeregisterPackageUse(PackageIdentity package, ServerTaskID serverTaskID)
+        {
+            try
+            {
+                var repository = repositoryFactory.CreateJournalRepository();
+
+                if (repository.TryGetJournalEntry(package, out var entry))
+                {
+                    entry.PackageLocks.RemoveLock(serverTaskID);
+                }
+            }
+            catch (Exception ex)
+            {
+                //We need to ensure that an issue with the journal doesn't interfere with the deployment.
+                log.Error($"Unable to deregister package use for retention.{Environment.NewLine}{ex.ToString()}");
+            }
+        }
+
+        public bool HasLock(PackageIdentity package)
+        {
+            return repositoryFactory.CreateJournalRepository()
+                                    .TryGetJournalEntry(package, out var entry)
+                   && entry.PackageLocks.HasLock();
+        }
+
+        public IEnumerable<DateTime> GetUsage(PackageIdentity package)
+        {
+            return repositoryFactory.CreateJournalRepository()
+                                    .TryGetJournalEntry(package, out var entry)
+                ? entry.PackageUsage.GetUsageDetails()
+                : new DateTime[0];
```

Comment:
`Array.Empty<DateTime>()`

Reply:
Unfortunately Calamari is Framework 4.0, which [doesn't have Array.Empty()](https://docs.microsoft.com/en-us/dotnet/api/system.array.empty?view=net-5.0)...

Reply:
😢